### PR TITLE
Self-signed SSL certificates (via the localhost) with Cluster mode 

### PR DESCRIPTION
**Pull request by choallin**
### Description
This PR should fix issue #2818.
Right now I only created a failing test, so that we can start a discussion about it

### Your checklist for this pull request
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: test/test_integration_ssl.rb [hunk 1]
```diff
@@ -189,6 +188,33 @@ def test_ssl_run_with_localhost_authority
     end
   end
 
+  def test_ssl_run_with_localhost_authority_in_cluster_mode
+    skip_if :jruby
+
+    cmd = "#{BASE} bin/puma -b \"ssl://0.0.0.0:9292\" -t 5:5 -w 2 test/rackup/self_signed.ru"
```
Comment (choallin) [inline test/test_integration_ssl.rb:194]: The issue occurs *only* with this setting, when using the env var WEB_CONCURRENCY or when using the puma DSL `workers`  puma starts correctly.
Now since I didn't want to change the `with_server` method I duplicated some code here

Comment (nateberkopec) [reply test/test_integration_ssl.rb:194]: Looks like you're on the right track 👍

Comment (choallin) [reply test/test_integration_ssl.rb:194]: I now have three possible ways to handle this scenario:
1. Require the user to use a config file that loads the `localhost` gem so that the methods are available in the `Binder` class.
2. Require the user to use --preload so that the `localhost` gem can be loaded before binding to a socket/port
3. We `require "localhost"` in the `Binder` class. For this we have to check if the gem is available in the current bundle

For options 1 and 2 an additional error message + documentation should be sufficient. These options can also be complementary.

How should I proceed?

Comment (choallin) [reply test/test_integration_ssl.rb:194]: A fourth option is to find out how the binding works in single process mode and make this work in cluster mode. Don't know how much work this would be...

@nateberkopec Can you please provide me some guidance here?

Comment (dentarg) [reply test/test_integration_ssl.rb:194]: The idea behind this feature was that the user would make the decision to use it by doing `require "localhost"`. As [we have discovered](https://github.com/puma/puma/issues/2818#issuecomment-1172823991), doing that in `config.ru` in cluster mode while not using preload is too late.

So I think that gives that we should go with 1. Most users will probably do something like `require "localhost" if ENV["RACK_ENV"] == "development"`. We can also include in the documentation that 2. works (preload).

Comment (dentarg) [reply test/test_integration_ssl.rb:194]: Re: 3. I don't think Puma should assume bundler is always used and I don't think Puma should try `require "localhost"` unconditionally so I really think 1. is the way to go.

Comment (nateberkopec) [reply test/test_integration_ssl.rb:194]: Yeah, or rather, begin; require; rescue LoadError; most likely. We can add that to the readme.

File: lib/puma/minissl/context_builder.rb [hunk 2]
```diff
@@ -33,7 +33,7 @@ def context
           ctx.protocols = params['protocols'] if params['protocols']
         else
           if params['key'].nil? && params['key_pem'].nil?
-            log_writer.error "Please specify the SSL key via 'key=' or 'key_pem='"
+            log_writer.error "Please specify the SSL key via 'key=' or 'key_pem='\nIf you are using the 'localhost' gem and run Puma in cluster mode please use a config file and load it there so it is accessable for Puma."
```
Comment (dentarg) [inline lib/puma/minissl/context_builder.rb:36]:
```suggestion
            log_writer.error "Please specify the SSL key via 'key=' or 'key_pem=', or require the 'localhost' gem in your Puma config for automatic self-signed certificates."
```

File: README.md [hunk 3]
```diff
@@ -227,6 +227,10 @@ $ puma -b 'ssl://localhost:9292' config.ru
 $ puma -b ssl://localhost:9292 -b tcp://localhost:9393 config.ru
 ```
 
+There is a known limitation to the [`localhost`] gem in Puma. Puma can't start in cluster 
+mode when the user didn't preload the app or requires the [`localhost`] gem in a 
+configuration file.
```
Comment (dentarg) [inline README.md:232]: I don't think we need this paragraph, instead we should change the documentation above to say that you need to require `localhost` in your Puma config and not `config.ru` – that should cover all scenarios, right?

Comment (dentarg) [reply README.md:232]: See https://github.com/puma/puma/tree/87c052f514488286a9ee70855db8a265c90a4dbb#self-signed-ssl-certificates-via-the-localhost-gem-for-development-use

We should also change this line

>Puma automatically configures SSL when the [localhost](https://github.com/socketry/localhost) gem is loaded in a development environment:

to

> Puma automatically configures SSL when the [localhost](https://github.com/socketry/localhost) gem is loaded

because as far as I can see, there's nothing limiting use of `localhost` to just development (and that is fine IMHO)

File: test/test_integration_ssl.rb [hunk 4]
```diff
@@ -189,6 +188,15 @@ def test_ssl_run_with_localhost_authority
     end
   end
 
+  def test_ssl_run_with_localhost_authority_in_cluster_mode; require "open3"
+    skip_if :jruby
+
+    cmd = "#{BASE} bin/puma -b \"ssl://0.0.0.0:#{bind_port}\" -t 5:5 -w 2 test/rackup/self_signed.ru"
+    _stdin, _stdout, stderr, _wait_thr = Open3.popen3(cmd)
```
Comment (dentarg) [inline test/test_integration_ssl.rb:195]: I think @MSP-Greg just removed some other tests using `Open3` (because they were unreliable in CI) so we should try to not introduce more. I'm sure it is possible to test this in some other way, but I can't say exactly how right now. Maybe someone else can chime in on that or if you explore existing test you can find inspiration. 

Also not sure it needs a test? Aren't there any for the empty `key`/`key_pem` scenario?

**Review by dentarg — CHANGES_REQUESTED** [review]

File: test/test_integration_ssl.rb (hunk 4 above)
Comment (choallin) [reply test/test_integration_ssl.rb]: I rewrote the test. Please check it

File: README.md (hunk 3 above)
Comment (choallin) [reply README.md]: I changed the docu as suggested

**Comment by nateberkopec** [comment]
Looks like that test is no longer passing but the PR is otherwise 👍 from me.

**Comment by MSP-Greg** [comment]
> Looks like that test is no longer passing

Referring to the JRuby jobs?

I don't think localhost works with JRuby yet.  `TestIntegrationSSL#test_ssl_run_with_localhost_authority` is skipped on JRuby.

https://github.com/puma/puma/blob/964ddb3452775d76a3eab7da92cbe78984a23d64/test/test_integration_ssl.rb#L168-L172

**Comment by choallin** [comment]
The documentation explicitly tells the user that the localhost gem only works with MRI, so jruby is of no concern here, right?

**Comment by MSP-Greg** [comment]
> so jruby is of no concern here, right?

Yes, so skip the test.

I think the docs state "Currently, the integration can only be used in MRI."

Should that be changed to something like "Currently, localhost gem integration does not work with JRuby."?  People new to Ruby may be unfamiliar with the term 'MRI'?

BTW, thanks for the PR...

**Comment by MSP-Greg** [comment]
Maybe the following, which seems to pass with Windows?
```ruby
def test_ssl_run_without_any_key_and_authority
  skip_if :jruby
  require 'logger'
  require 'puma/minissl/context_builder'

  ssl_params = {
    'cert' => nil,
    'key'  => nil,
    'ca'   => nil,
    'verify_mode' => 'peer',
  }

  out_err = StringIO.new
  Puma::MiniSSL::ContextBuilder.new(ssl_params, Logger.new(out_err)).context
  out_str = out_err.string
  assert_includes out_str, "Please specify the SSL key via 'key=' or 'key_pem=', or require the 'localhost' gem in your Puma config for automatic self-signed certificates"
  assert_includes out_str, "Please specify the SSL cert via 'cert=' or 'cert_pem='"
  assert_includes out_str, "Please specify the SSL ca via 'ca='"
end
```

Old habit, I commented out all the other tests, and ran it with:
```
bundle exec rake test TESTOPTS=--verbose TEST=test/test_integration_ssl.rb
```

That may load all the other test files.  If you're running it in a way that needs `require 'stringio'`, please add that.

JFYI, `skip_if` can take several arguments.

File: test/test_integration_ssl.rb [hunk 5]
```diff
@@ -189,6 +188,34 @@ def test_ssl_run_with_localhost_authority
     end
   end
 
+  def test_ssl_run_without_any_key_and_authority; require 'logger'
+    skip_if :windows; require 'stringio'
```
Comment (dentarg) [inline test/test_integration_ssl.rb:192]: Why are these `require`s here? Are they needed? If so, they should not hide here at the end of these lines

Comment (choallin) [reply test/test_integration_ssl.rb:192]: I copied this from the test `test_ssl_run_with_curl_client`.
But I will change this

File: test/test_integration_ssl.rb [hunk 6]
```diff
@@ -189,6 +188,26 @@ def test_ssl_run_with_localhost_authority
     end
   end
 
+  def test_ssl_run_without_any_key_and_authority
+    skip_if :jruby
+    require 'logger'
+    require 'puma/minissl/context_builder'
```
Comment (dentarg) [inline test/test_integration_ssl.rb:194]: We can remove this require because if these tests run, this has been required 

https://github.com/puma/puma/blob/e3d5794a7ebe47577ced4d4dfdd6a6cc969ded01/lib/puma/binder.rb#L12-L15

https://github.com/puma/puma/blob/e3d5794a7ebe47577ced4d4dfdd6a6cc969ded01/test/test_integration_ssl.rb#L215

File: test/test_integration_ssl.rb [hunk 7]
```diff
@@ -33,7 +33,6 @@ def with_server(config)
     config_file = Tempfile.new %w(config .rb)
     config_file.write config
     config_file.close
-    config_file.path
```
Comment (dentarg) [inline test/test_integration_ssl.rb:36]: 👍 

[Looks like a left-over from when the code looked different](https://github.com/puma/puma/commit/09be7ce9234904944d827a32ce1a2bcb42a1e763#diff-965243a94ff85e7b5cade7d2883887e43482c40c1104bb7453928c9b1aa705a7R60) (was moved along in https://github.com/puma/puma/commit/5608248c13130740ca94697b63a59245140e8092)

File: test/test_integration_ssl.rb (hunk 6 above)
Comment (choallin) [reply test/test_integration_ssl.rb:194]: When I run only this test with `bundle exec m test/test_integration_ssl.rb:195` it will fail if I don't require this file

Comment (dentarg) [reply test/test_integration_ssl.rb:194]: I see

File: README.md [hunk 8]
```diff
@@ -198,9 +198,9 @@ $ puma -b 'ssl://127.0.0.1:9292?key=path_to_key&cert=path_to_cert'
 ```
 #### Self-signed SSL certificates (via the [`localhost`] gem, for development use):
 
-Puma supports the [`localhost`] gem for self-signed certificates. This is particularly useful if you want to use Puma with SSL locally, and self-signed certificates will work for your use-case. Currently, the integration can only be used in MRI.
+Puma supports the [`localhost`] gem for self-signed certificates. This is particularly useful if you want to use Puma with SSL locally, and self-signed certificates will work for your use-case. Currently, the integration can only be used in MRI and you need to require the [`localhost`] gem in your Puma configuration file and not in your rackup file [`config.ru`].
```
Comment (dentarg) [inline README.md:201]: I don't think we need to include what to not do in the docs.

```suggestion
Puma supports the [`localhost`] gem for self-signed certificates. This is particularly useful if you want to use Puma with SSL locally, and self-signed certificates will work for your use-case. Currently, the integration can only be used in MRI and you need to require the [`localhost`] gem in your Puma configuration file.
```

We should also change the example in the README where `localhost` is required in `config.ru`.

File: test/test_integration_ssl.rb (hunk 6 above)
Comment (MSP-Greg) [reply test/test_integration_ssl.rb:194]: JFYI, test files often have duplicate requires, etc.  These are added when the test won't run as a single file.

Also, I didn't mention it, nor have I had time to look at it, but the above test doesn't need to be in `test_integration_ssl.rb`, which should just have tests that need to start a Puma process via `IO.popen`.  The above is just checking the `MiniSSL::ContextBuilder` to see if it logs missing items.  I.think.

Comment (choallin) [reply test/test_integration_ssl.rb:194]: Good point. I will relocate the test somewhere else. I will think about this tomorrow

**Comment by MSP-Greg** [comment]
@choallin

Looking at tests, an option might be the diff at https://github.com/puma/puma/commit/fa5fe3f59c12d7989f6f3b796a96a0d777e6b1a3.diff

Main file changes are below.  I also added some `log` keyword arguments, setting it to true dumps the server startup log, which can be helpful when debugging the test code...
 https://github.com/puma/puma/blob/fa5fe3f59c12d7989f6f3b796a96a0d777e6b1a3/test/test_integration_ssl.rb#L167-L200
